src/bun.js/bindings/JSBuffer.cpp

```diff
@@ -993,120 +1006,111 @@ static inline JSC::EncodedJSValue jsBufferPrototypeFunction_fillBody(JSC::JSGlob
-            if (!encoded) {
-                throwTypeError(lexicalGlobalObject, scope, "Invalid encoding"_s);
-                return JSC::JSValue::encode(jsUndefined());
-            }
+    if (UNLIKELY(end > limit)) {
+        throwRangeError(lexicalGlobalObject, scope, "end out of range"_s);
+        return JSC::JSValue::encode(jsUndefined());
+    }
 
-            encoding = encoded.value();
-        }
+    if (value.isString()) {
+        auto startPtr = castedThis->typedVector() + start;
+        auto str_ = value.toWTFString(lexicalGlobalObject);
+        ZigString str = Zig::toZigString(str_);
 
-        if (!offsetValue.isUndefined()) {
-            if (auto offset = offsetValue.tryGetAsUint32Index()) {
-                start = offset.value();
-            } else {
-                throwVMError(lexicalGlobalObject, scope, createRangeError(lexicalGlobalObject, "start out of range"_s));
-                return JSC::JSValue::encode(jsUndefined());
-            }
+        if (str.len == 0) {
+            memset(startPtr, 0, end - start);
+        } else if (UNLIKELY(!Bun__Buffer_fill(&str, startPtr, end - start, encoding))) {
+            throwTypeError(lexicalGlobalObject, scope, "Failed to decode value"_s);
+            return JSC::JSValue::encode(jsUndefined());
         }
+    } else if (auto* view = JSC::jsDynamicCast<JSC::JSUint8Array*>(value)) {
+        auto* startPtr = castedThis->typedVector() + start;
+        auto* head = startPtr;
+        size_t remain = end - start;
 
-        if (!lengthValue.isUndefined()) {
-            if (auto length = lengthValue.tryGetAsUint32Index()) {
-                end = std::min(static_cast<size_t>(length.value()), end);
-            } else {
-                throwVMError(lexicalGlobalObject, scope, createRangeError(lexicalGlobalObject, "end out of range"_s));
-                return JSC::JSValue::encode(jsUndefined());
-            }
+        if (UNLIKELY(view->isDetached())) {
+            throwVMTypeError(lexicalGlobalObject, scope, "Uint8Array is detached"_s);
+            return JSValue::encode(jsUndefined());
         }
 
-        if (start >= end) {
-            return JSValue::encode(castedThis);
+        size_t length = view->byteLength();
+        if (UNLIKELY(length == 0)) {
+            throwTypeError(lexicalGlobalObject, scope, "Buffer cannot be empty"_s);
+            return JSC::JSValue::encode(jsUndefined());
         }
 
-        auto startPtr = castedThis->typedVector() + start;
-        auto str_ = value.toWTFString(lexicalGlobalObject);
-        ZigString str = Zig::toZigString(str_);
-
-        if (str.len > 0) {
-            Bun__Buffer_fill(&str, startPtr, end - start, encoding);
-        } else {
-            memset(startPtr, 0, end - start);
+        memcpy(head, view->typedVector(), length);
+        remain -= length;
+        head += length;
+        while (remain >= length) {
+            memcpy(head, startPtr, length);
```

Comment:
can we use memmove instead of memcpy 

just to avoid the potential UB from overlalping buffers

Reply:
Good point - I should go comb over other calls within this file too.

Reply:
Done in https://github.com/oven-sh/bun/pull/2341/commits/e0c5c4aa3a60d39e214b8103ca998744221130fc in two places &minus; this particular instance copies within itself, which is guaranteed not to overlap.